### PR TITLE
[Security] Add the isEqualTo method implementation

### PR DESCRIPTION
**Pull request by Guikingone**
Hi, 

With the logout_on_user_change key addition in the security.yaml, every User entity must implement the ``isEqualTo`` method from the ``UserInterface`` .

### CODE REVIEW

**Comment by javiereguiluz** [comment]
Not sure about this change. We've updated for example the Symfony Demo app to 4.0 without having to do this change. I'm not saying that your proposal is wrong, but I'd like to read more comments about this. Thanks!

File: security/entity_provider.rst [hunk 1]
```diff
@@ -140,6 +140,40 @@ For this entry, suppose that you already have a ``User`` entity inside an
...
+     * @ORM\Table(name="app_users")
+     * @ORM\Entity(repositoryClass="AppBundle\Repository\UserRepository")
+     */
+    class User implements UserInterface, EquatableInterface, \Serializable
+    {
+        // ...
+        
+        public function isEqualTo(UserInterface $user) 
+        {
+            if ($this->username !== $user->getUsername() {
```
Comment (apfelbox) [inline security/entity_provider.rst]:
```php
return $this->username === $user->getUsername();
```

?

**Comment by Guikingone** [comment]
@javiereguiluz Yep, I've seen that the demo has been updated and that this interface isn't implemented, for me, the problem occurs when I use the ``logout_on_user_change`` key inside the security.yaml.

File: security/entity_provider.rst (hunk 1 above)
Comment (Guikingone) [reply security/entity_provider.rst]: Yep, work fine, I wasn't sure about the capacity of the security layer to recognize the boolean after the condition.

**Comment by xabbuh** [comment]
The proposed implementation does not take the password into account (which was the main reason IIRC to introduce this option at all). I would not advise to promote it as a default.

**Comment by javiereguiluz** [comment]
I still don't understand this change ... maybe @chalasr can shed some light here. Thanks!

**Comment by Guikingone** [comment]
@javiereguiluz As far as I can see from now (in fact, as I have launched 4 projects using Symfony 4. *), if the ``EquatableInterface`` is not implemented, the User is logged out every time it moves from one page to another.

**Comment by xabbuh** [comment]
@Guikingone This only happens if the refreshed user does not match the user object that is stored in the session. From my experience, this is often caused by removing the hashed password from the user before passing it to the token. But IMO that's not a solution we should recommend and thus implementing the `EquatableInterface` is not something that is mandatory.

**Comment by javiereguiluz** [comment]
@Guikingone in the Symfony Demo app we don't implement the `EquatableInterface` and we don't suffer from the error you mention. The reason is probably what @xabbuh mentioned. So, what should we do with this PR? Close it as "won't merge"? Reword it to explain this as a rare edge case? Thanks!

**Comment by Guikingone** [comment]
@javiereguiluz I think it can be useful to mention this case but as @xabbuh mention it, it's a really rare case and I don't want to overload the documentation for a "rare" case. 

Maybe mention the fact that if the user password isn't in the one which is stored in the security token, using the ``EquatableInterface`` can be a solution?

**Comment by xabbuh** [comment]
I would then instead just explain how two user objects are compared and after that explain that you can implement this interface if you need some logic that differs from the default.

**Comment by Guikingone** [comment]
@xabbuh Yep, can be a good idea, I'm gonna look at this approach this weekend.

**Comment by Guikingone** [comment]
@javiereguiluz Yes, sorry for the delay, I look at the proposal and the implementation this week, I've forgotten this issue (along with the #9513 one), really sorry.

**Comment by javiereguiluz** [comment]
@Guikingone I'm sorry but I must close this pull request as "won't fix". I appreciate the work you did here, but things related to security are going to change a lot.

We've decided that "enough is enough" and we're going to revamp all docs to drastically simplify things. For example, this will be the tiny mention to `EquatableInterface ` in the new docs: https://github.com/symfony/symfony-docs/pull/10423/files#diff-feb08062dab1914c2f5ecbae1f547ddbR176

I hope you understand it. Thanks!
